comparison/scripts/nbconverted/0.get-cytominer-tool-differences.py

```diff
@@ -0,0 +1,544 @@
+# * Plate
+# 
+# Other metadata columns are not expected to be aligned.
+# For example, we have [updated MOA and Target information](https://github.com/broadinstitute/lincs-cell-painting/issues/11) in the pycytominer version.
+# 
+# ## Data Levels
+# 
+# Image-based profiling results in the following output data levels.
+# We do not compare all data levels in this notebook.
+# 
+# | Data | Level | Comparison |
+# | :---- | :---- | :-------- |
+# | Images | Level 1 | NA |
+# | SQLite File (single cell profiles ) | Level 2 | NA |
+# | Aggregated Profiles with Well Information (metadata) | Level 3 | Yes  |
+# | Normalized Aggregated Profiles with Metadata | Level 4a | Yes | 
+# | Normalized and Feature Selected Aggregated Profiles with Metadata | Level 4b | Yes |
+# | Perturbation Profiles created Summarizing Replicates  | Level 5 | No |
+
+# In[1]:
+
+
+get_ipython().run_line_magic('load_ext', 'nb_black')
+
+
+# In[2]:
+
+
+import os
+import pathlib
+import numpy as np
+import pandas as pd
+
+from util import build_file_dictionary, load_data, generate_output_filenames
+
+
+# In[3]:
+
+
+# Pycytominer plates are saved with 5 floating point decimals
+round_decimals = 5
+
+# Create the output directory
+output_dir = "results"
+os.makedirs(output_dir, exist_ok=True)
+
+
+# In[4]:
+
+
+def get_metrics(pycyto_df, cyto_df, features):
+    # Align features
+    pycyto_df = pycyto_df.reindex(features, axis="columns")
+    cyto_df = cyto_df.reindex(features, axis="columns")
+
+    # Assess difference
+    abs_diff = pycyto_df.subtract(cyto_df).abs()
+    mean_diff = abs_diff.mean()
+    median_diff = abs_diff.median()
+    sum_diff = abs_diff.sum()
+
+    complete_mean_diff = mean_diff.replace([np.inf, -np.inf], np.nan).dropna().mean()
+    complete_median_diff = (
+        median_diff.replace([np.inf, -np.inf], np.nan).dropna().mean()
+    )
+    complete_sum_diff = sum_diff.replace([np.inf, -np.inf], np.nan).dropna().sum()
+
+    return (
+        mean_diff,
+        complete_mean_diff,
+        median_diff,
+        complete_median_diff,
+        sum_diff,
+        complete_sum_diff,
+    )
+
+
+def find_feature_diff(pycyto_df, cyto_df, plate, all_features):
+    all_features_df = pd.DataFrame(
+        ["missing"] * len(all_features), index=all_features, columns=[plate]
+    )
+    pycyto_features = set(pycyto_df.columns.tolist())
+    cyto_features = set(cyto_df.columns.tolist())
+    present_both = pycyto_features.intersection(cyto_features)
+
+    all_features_df.loc[
+        all_features_df.index.isin(pycyto_features), plate
+    ] = "only_pycytominer"
+    all_features_df.loc[
+        all_features_df.index.isin(cyto_features), plate
+    ] = "only_cytominer"
+    all_features_df.loc[
+        all_features_df.index.isin(present_both), plate
+    ] = "present_both"
+
+    return all_features_df
+
+
+# In[5]:
+
+
+pycytominer_dir = pathlib.Path("../profiles/backend/")
+cytominer_dir = pathlib.Path(
+    "/Users/gway/work/projects/"
```

Comment:
For the purpose of reproducibility would it be better to not hardcode the path?

Reply:
absolutely better, I agree. However, there are limitations to what we can do to improve the current cytominer profile data reproducibility. I think the best thing that we can do now (to both minimize time and maximize reproducibility) is to add documentation to why we need to hardcode this path.

As it currently stands and to my knowledge, there is no way for anyone outside the lab (i.e. without access to the imaging-platform S3 bucket) to reproduce this script. There are many more things we need to do in order to enable this - all of which are outside the scope of this PR.

Reply:
comment added in 4a2edbce50a93e24eb16cb99ebbff4c6de6a54c1

Reply:
Agreed 👍. You may consider including instructions like [this](https://github.com/broadinstitute/2018_02_07_Image_based_Profiling_Pilot/blob/master/README.md#step-3-create-directories-and-download-files-from-aws) provided you think it will take <15 mins to do so.

Reply:
> (and use Path.home()).

Oh, this is great - I will update to use this.

> You may consider including instructions like this

I think linking to a publicly available repo with these steps is a good idea. Otherwise, I think adding these instructions is too much detail. Also, there are many ways of updating this path for it to work. Lastly, there are checks inside this script that will throw errors if plates are not aligned - i.e. it would be tough to add an incorrect path here and have the script to silently fail. It should fail loudly with an incorrect path!

Reply:
`Path.home()` used in 76d2db34252738735ff00392034f2e429ff8d383

wooo! reduced hardcoding 🎉